### PR TITLE
Encode x86 SIMD instructions with REX prefix; fixes #1090

### LINKED ISSUE #1090: Encode x86 SIMD instructions with REX prefix to use more XMM registers
See discussion https://github.com/CraneStation/cranelift/pull/992#discussion_r325063181

### PR DESCRIPTION
**Pull request by abrown**
As a preliminary step to this, I re-factored `PerCpuModeEncodings` to reduce some of the issues observed in #955. Although this was not strictly necessary, I felt that changing many of the SIMD encodings was a good time to take a look at `PerCpuModeEncodings`. Stepping through the commits one-by-one should show the re-factoring progression. The final step, adding the REX prefix, should be more clear in the code after the re-factoring.

### CODE REVIEW

File: cranelift-codegen/meta/src/isa/x86/encodings.rs [hunk 1]
```diff
@@ -22,6 +22,10 @@ use super::recipes::{RecipeGroup, Template};
 use crate::cdsl::formats::FormatRegistry;
 use crate::cdsl::instructions::BindParameter::Any;
 
+enum EncodingPattern {
```
Comment (bnjbvr) [inline cranelift-codegen/meta/src/isa/x86/encodings.rs]: How do you feel about naming this CpuMode? (even if there's already something named this way somewhere else, I don't think the two namings would interact or conflict in any way)

File: cranelift-codegen/meta/src/isa/x86/encodings.rs [hunk 2]
```diff
@@ -41,6 +45,32 @@ impl<'defs> PerCpuModeEncodings<'defs> {
...
+        &self,
+        inst: impl Into<InstSpec>,
+        recipe: EncodingRecipe,
+        bits: u16,
+    ) -> EncodingBuilder {
+        EncodingBuilder::new(inst.into(), recipe.clone(), bits, self.formats)
+    }
+
+    /// Create an encoding builder from a template.
+    fn enc(&self, inst: impl Into<InstSpec>, template: Template) -> EncodingBuilder {
```
Comment (bnjbvr) [inline cranelift-codegen/meta/src/isa/x86/encodings.rs]: Is the idea to remove the other helper functions, and just use a combination of these new ones instead?

Comment (abrown) [reply cranelift-codegen/meta/src/isa/x86/encodings.rs]: Yes! For me it was initially very confusing to figure out what was going on in this file, if/should I create new helper methods, figuring out if a helper method I needed already existed, etc. I think having fewer methods (`add` and `enc`) would have helped me figure out more clearly what was going on.

**Comment by sstangl** [comment]
Just as a note, for #1101 , I'm going to try having a single Encoding for REX/non-REX. Hopefully that can be applied to SIMD instructions too.

I think the vague team consensus is to move away from Recipes as much as possible.

**Pull request closed without merging**